### PR TITLE
[MPS] Fix cat int32 offset overflow for outputs over 2^31 elements

### LINKED ISSUE #189960: [MPS] torch.cat silently corrupts output and writes out of bounds when the output exceeds 2^31 elements
### 🐛 Describe the bug

`torch.cat` on MPS returns silently wrong results when the output has more than
2^31 elements but every individual dimension is smaller than 2^31. Part of the
output is never written, and the kernel also writes out of bounds *before* the
output allocation, so it can corrupt unrelated live MPS tensors.

```python
import torch

rows, cols_half = 32769, 32768  # output (32769, 65536): numel = 2^31 + 65536
a = torch.full((rows, cols_half), 1, dtype=torch.int8, device="mps")
b = torch.full((rows, cols_half), 2, dtype=torch.int8, device="mps")
out = torch.cat([a, b], dim=1)
torch.mps.synchronize()

last = out[-1].cpu()
wrong = int((last[:cols_half] != 1).sum()) + int((last[cols_half:] != 2).sum())
boundary_row = (2**31) // (2 * cols_half)  # first row whose linear offsets cross 2^31
print(f"numel = 2^31 + {out.numel() - 2**31}")
print(f"wrong entries in last row: {wrong} / {2 * cols_half}")
print(f"unique values in first row past 2^31 (expected [1, 2]): {out[boundary_row].cpu().unique().tolist()}")
```

Output on 2.13.0:

```text
numel = 2^31 + 65536
wrong entries in last row: 65536 / 65536
unique values in first row past 2^31 (expected [1, 2]): [0]
```

Expected behavior: every row of `out` is 32768 ones followed by 32768 twos. A
control at output numel exactly 2^31 (`rows = 32768`) passes; the corruption
starts with the first element whose linear output offset reaches 2^31.

Also reproduced on nightly 2.14.0.dev20260710. The repro peaks at about 4.3 GB
of MPS allocations, and because the failing kernel writes at wrapped negative
offsets relative to the output buffer, it should be run in a process with no
other live MPS workload.

`cat_out_mps` selects the int32 kernel specialization based on
`isTooLargeForMPSGraph`, which checks individual dimension sizes but never the
total element count:
https://github.com/pytorch/pytorch/blob/a57319c678176efe5b0948ee878db668d3b7a20a/aten/src/ATen/native/mps/operations/Shape.mm#L174-L184
The Metal kernel accumulates the linear output offset in the 32-bit index type,
which wraps negative once the offset reaches 2^31:
https://github.com/pytorch/pytorch/blob/a57319c678176efe5b0948ee878db668d3b7a20a/aten/src/ATen/native/mps/kernels/Shape.metal#L27-L41
An int64 specialization already exists and is selected for the
`has_large_tensor` case.

Note that #188200 (perf rework for #188198) rewrites the contiguous same-dtype
fast path and may incidentally avoid this repro if it lands, but noncontiguous
or mixed-dtype concatenations would still take `cat_out_mps_impl<int32_t>`, and
there is currently no >= 2^31 correctness regression test for `cat` either way.

### Versions

```text
PyTorch version: 2.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5.2 (arm64)
Clang version: 21.0.0 (clang-2100.0.123.102)

Python version: 3.12.13 (main, May 10 2026, 19:20:41) [Clang 22.1.3 ] (64-bit runtime)
Python platform: macOS-26.5.2-arm64-arm-64bit

CPU:
Apple M5 Max
```

Also reproduced on torch==2.14.0.dev20260710.

cc @ezyang @gchanan @kadeng @msaroufim @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by bogatyy**
The Metal cat kernel accumulates linear input/output offsets as `sum(stride[dim] * dim_idx)` in its index type, but `cat_out_mps` selected the int32 specialization via `isTooLargeForMPSGraph`, which only checks individual dimension sizes. A contiguous output with more than 2^31 total elements but every dimension below 2^31 therefore wrapped `output_offset` negative: out-of-bounds GPU writes before the output allocation and a never-written tail.

Select the specialization from the largest linear offset each tensor can produce instead. Compared to checking `out.numel()` alone, this costs the same and additionally covers non-dense layouts (e.g. strided `out=` views) whose max offset exceeds their numel.

The new `test_64bit_cat` reproduces the issue's shape (`(32769, 65536)` int8, `dim=1`, numel = 2^31 + 65536) and checks rows on both sides of the 2^31 boundary plus the final row; it needs ~4.3 GB and is gated at 8 GB like the neighboring `TestLargeTensors` cases. Existing cat tests (198) pass unchanged.

Fixes #189960

*This PR was authored with assistance from Claude.*

### CODE REVIEW

**Comment by bogatyy** [comment]
@malfet gentle ping.

I kept the regression test in test_mps because the bug is specific to the MPS cat kernel's index-width selection. CPU cat uses 64-bit TensorIterator and CUDA gates on canUse32BitIndexMath, so neither has this failure mode, and the device-generic cat test class (TestTensorCreation) is not MPS-instantiated. It is also a large-memory test (needs ~4GB, gated on recommended_max_memory), the same category as test_linear_large. The existing device-generic test_cat_large_tensor covers a different regime (a 1D cat whose output stays under 2^31), so it would not have caught this.

File: aten/src/ATen/native/mps/operations/Shape.mm [hunk 1]
```diff
@@ -171,10 +185,10 @@ static void cat_out_mps_impl(const ITensorListRef& inputs, int64_t dimension, co
   }
 
   auto materialized_inputs = inputs.materialize();
-  bool has_large_tensor =
-      isTooLargeForMPSGraph(out) || std::any_of(materialized_inputs.begin(), materialized_inputs.end(), [](auto& t) {
-        return !cat_should_skip_tensor(t) && isTooLargeForMPSGraph(t);
-      });
+  bool has_large_tensor = needs_64bit_indexing(out) || isTooLargeForMPSGraph(out) ||
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/operations/Shape.mm]: It's not ideal to iterate through the sizes twice. Maybe instead of adding a separate `needs_64bit_indexing` function, we should add an optional flag to `isTooLargeForMPSGraph` which will turn on the `sum(stride * idx)` check, so that we can do both checks at the same time while only iterating through the sizes once

**Review by kurtamohler — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: test/test_mps.py [hunk 2]
```diff
@@ -10785,6 +10785,27 @@ def test_64bit_index_copy(self):
...
         torch.mps.empty_cache()
 
+    @serialTest()
+    def test_64bit_cat(self):
+        # https://github.com/pytorch/pytorch/issues/189960
+        # cat chose its int32 kernel from per-dim sizes alone, so an output
+        # with numel > 2**31 but every dim < 2**31 wrapped the kernel's linear
+        # output offset, writing out of bounds and leaving the tail stale.
+        if torch.mps.recommended_max_memory() < 8_000_000_000:
+            raise unittest.SkipTest("Needs at least 8Gb of RAM")
```
Comment (kurtamohler) [inline test/test_mps.py]: Probably should use `@largeTensorTest("8GB", device='mps')` instead

**Comment by bogatyy** [comment]
@kurtamohler Thank you for the review! PTAL, I have addressed both comments and force-pushed the fix (via amend).

**Review by kurtamohler — APPROVED** [review]

**Comment by chao121026** [comment]
Hello, I can confirm this bug on both **noncontiguous** and **mixed-dtype** 
paths, on a newer nightly than the reported version.

### Full environment and test output
======================================================================
ENVIRONMENT
======================================================================
PyTorch version:      2.15.0.dev20260915
PyTorch git version:  19435f3e1372ffcedf95125df1524e53f4cd26cf
Python version:       3.11.16
Platform:             macOS-15.7.8-arm64-arm-64bit
Machine:              arm64

macOS version:
ProductName:		macOS
ProductVersion:		15.7.8
BuildVersion:		24G824

Chip:                 Apple M4 Pro
Physical memory:      24.0 GB

MPS available:        True
MPS built:            True
======================================================================

[Path 1] Contiguous + same dtype
----------------------------------------------------------------------
a.is_contiguous(): True
b.is_contiguous(): True
out.dtype: torch.int8
out.is_contiguous(): True
numel = 2^31 + 65536
wrong entries: 0 / 65536
unique values past 2^31 (expected [1, 2]): [1, 2]

[Path 2] Noncontiguous + same dtype
----------------------------------------------------------------------
a.is_contiguous(): False
a.shape: torch.Size([32769, 32768])
a.stride(): (1, 32769)
b.is_contiguous(): True
out.dtype: torch.int8
out.is_contiguous(): True
numel = 2^31 + 65536
wrong entries: 65536 / 65536
unique values past 2^31 (expected [1, 2]): [0]

[Path 3] Contiguous + mixed dtype (int8 + int16)
----------------------------------------------------------------------
a.dtype: torch.int8
b.dtype: torch.int16
a.is_contiguous(): True
b.is_contiguous(): True
out.dtype: torch.int16
out.is_contiguous(): True
numel = 2^31 + 65536
wrong entries: 65536 / 65536
unique values past 2^31 (expected [1, 2]): [0]

======================================================================


So the contiguous same-dtype case is fixed, but noncontiguous 
and mixed-dtype still fail. Both of those go through 
`cat_out_mps_impl<int32_t>`, which is selected by 
`isTooLargeForMPSGraph` — and that function only checks 
per-dimension sizes, never `out.numel()`.

Also worth noting: I'm on M4 Pro, not M5 Max, so this isn't 
tied to one chip generation.

Would gating on `out.numel() > INT_MAX` cover all three paths? 
If you'd like, I can take a crack at the fix — let me know 
if that's useful.

**Comment by bogatyy** [comment]
@kurtamohler i've rebased against latest main, and also moved all tests from test_mps.py (which I learned is an anti-pattern) into test_tensor_creation_ops.py

**Comment by kurtamohler** [comment]
> moved all tests from test_mps.py (which I learned is an anti-pattern) into test_tensor_creation_ops.py

Honestly I think it would have been better to leave it where it was. I don't think we need it to run on all other platforms, since it's a relatively expensive test and it's specifically testing a limitation of the MPS platform

**Comment by bogatyy** [comment]
@kurtamohler sure, i reverted to the original version (test_mps), still freshly rebased
